### PR TITLE
Implement movsq

### PR DESCRIPTION
**Pull request by oureveryday**
Implement movsq I encountered in sprotect

### CODE REVIEW

**Review by NaC-L — COMMENTED** [review]
I havent reviewed it properly, but we should be able to have one movsX function to do all movs, movsb, movsw... etc. we also should be able to get the size info from operands. That would be a more elegant solution

**Comment by oureveryday** [comment]
I think maybe we have to pass a size paramater into that movsX?

**Comment by NaC-L** [comment]
I am not exactly sure, but operands[1].size should give us the correct size, so we dont really need to pass a parameter.

**Comment by oureveryday** [comment]
Done

File: lifter/Semantics.cpp [hunk 1]
```diff
@@ -259,33 +259,29 @@ void lifterClass::branchHelper(Value* condition, const string& instname,
...
-  // sign = (x*(x+1)) - 1
-  // v = sign * bytesize ; bytesize is 1
 
-  Value* Direction =
-      createSubFolder(createMulFolder(DF, createAddFolder(DF, one)), one);
+  // sign = (DF*(DF+1)) - 1
+  // v = sign * byteSize
+  Value* Direction = createMulFolder(
+      createSubFolder(createMulFolder(DF, createAddFolder(DF, one)), one),
+      byteSize);
```
Comment (NaC-L) [inline lifter/Semantics.cpp:284]: Now that I look back, we can just use a select here, 
if DF == 0, direction is -1, 
if DF == 1, direction is  1

File: lifter/Semantics.cpp [hunk 2]
```diff
@@ -3670,10 +3653,21 @@ void lifterClass::liftInstructionSemantics() {
...
+  case ZYDIS_MNEMONIC_MOVSW: {
+    lift_movsX();
+    break;
+  }
+  case ZYDIS_MNEMONIC_MOVSD: {
+    lift_movsX();
+    break;
+  }
+  case ZYDIS_MNEMONIC_MOVSQ: {
+    lift_movsX();
```
Comment (NaC-L) [inline lifter/Semantics.cpp:3668]: we can make the cases fall through such as

```
  case ZYDIS_MNEMONIC_MOVSB:
  case ZYDIS_MNEMONIC_MOVSW:
  case ZYDIS_MNEMONIC_MOVSD:
  case ZYDIS_MNEMONIC_MOVSQ: {
    lift_movsX();
    break;
  }
  ```
  
  would look more elegant

File: lifter/Semantics.cpp (hunk 1 above)
Comment (oureveryday) [reply lifter/Semantics.cpp:284]: > Now that I look back, we can just use a select here, if DF == 0, direction is -1, if DF == 1, direction is 1

I tried with 
```cpp
  Value* Direction =
      createSelectFolder(DF, ConstantInt::get(DF->getType(), 1),
        ConstantInt::get(DF->getType(), -1));
```

But it seems does something wrong.

File: lifter/Semantics.cpp (hunk 2 above)
Comment (NaC-L) [reply lifter/Semantics.cpp]: Thanks!

File: lifter/Semantics.cpp (hunk 1 above)
Comment (NaC-L) [reply lifter/Semantics.cpp:284]: Uhhh, I just realized, why are we doing `DF->getType()` ? Also should be 
```
 Value* Direction =
      createSelectFolder(DF, ConstantInt::get(__directionsize__, 1*bytesize),
        ConstantInt::get(__directionsize__, -1*bytesize));
        ```

Comment (oureveryday) [reply lifter/Semantics.cpp]: That worked.

**Review by NaC-L — APPROVED** [review]
LGTM

**Comment by NaC-L** [comment]
thank you for your contribution

**Pull request merged**